pkg/mei/mkhiclient.go

```diff
@@ -0,0 +1,309 @@
+	"MEI_DEV_ID_PBG_1":      0x1D3A,
+	"MEI_DEV_ID_PPT_1":      0x1E3A,
+	"MEI_DEV_ID_PPT_2":      0x1CBA,
+	"MEI_DEV_ID_PPT_3":      0x1DBA,
+	"MEI_DEV_ID_LPT_H":      0x8C3A,
+	"MEI_DEV_ID_LPT_W":      0x8D3A,
+	"MEI_DEV_ID_LPT_LP":     0x9C3A,
+	"MEI_DEV_ID_LPT_HR":     0x8CBA,
+	"MEI_DEV_ID_WPT_LP":     0x9CBA,
+	"MEI_DEV_ID_WPT_LP_2":   0x9CBB,
+	"MEI_DEV_ID_SPT":        0x9D3A,
+	"MEI_DEV_ID_SPT_2":      0x9D3B,
+	"MEI_DEV_ID_SPT_H":      0xA13A,
+	"MEI_DEV_ID_SPT_H_2":    0xA13B,
+	"MEI_DEV_ID_LBG":        0xA1BA,
+	"MEI_DEV_ID_BXT_M":      0x1A9A,
+	"MEI_DEV_ID_APL_I":      0x5A9A,
+	"MEI_DEV_ID_DNV_IE":     0x19E5,
+	"MEI_DEV_ID_GLK":        0x319A,
+	"MEI_DEV_ID_KBP":        0xA2BA,
+	"MEI_DEV_ID_KBP_2":      0xA2BB,
+	"MEI_DEV_ID_CNP_LP":     0x9DE0,
+	"MEI_DEV_ID_CNP_LP_4":   0x9DE4,
+	"MEI_DEV_ID_CNP_H":      0xA360,
+	"MEI_DEV_ID_CNP_H_4":    0xA364,
+	"MEI_DEV_ID_CMP_LP":     0x02e0,
+	"MEI_DEV_ID_CMP_LP_3":   0x02e4,
+	"MEI_DEV_ID_CMP_V":      0xA3BA,
+	"MEI_DEV_ID_CMP_H":      0x06e0,
+	"MEI_DEV_ID_CMP_H_3":    0x06e4,
+	"MEI_DEV_ID_CDF":        0x18D3,
+	"MEI_DEV_ID_ICP_LP":     0x34E0,
+	"MEI_DEV_ID_JSP_N":      0x4DE0,
+	"MEI_DEV_ID_TGP_LP":     0xA0E0,
+	"MEI_DEV_ID_MCC":        0x4B70,
+	"MEI_DEV_ID_MCC_4":      0x4B75,
+}
+
+// various ME-related constants
+const (
+	// ME - current working state set to normal
+	MEHfs1CwsNormal = 0x5
+	// ME - current operation mode set to normal
+	MEHfs1ComNormal = 0x0
+	// ME - CSE's firmware SKU is custom
+	MEHfs3FwSkuCustom = 0x5
+)
+
+// MKHIClient is a client to send MKHI commands via MEI.
+type MKHIClient struct {
+	MEI              MEI
+	clientProperties *ClientProperties
+}
+
+// MKHI command groups, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L22
+const (
+	MKHIGroupIDCbm    = 0x0
+	MKHIGroupIDHmrfpo = 0x5
+	MKHIGroupIDGen    = 0xff
+)
+
+// MKHI HMRFPO command IDs, see
+// https://github.com/coreboot/coreboot/blob/b8b8ec832360ada5a313f10938bb6cfc310a11eb/src/soc/intel/common/block/include/intelblocks/cse.h#L33
+const (
+	MKHIHmrfpoEnable    = 0x1
+	MKHIHmrfpoGetStatus = 0x3
+)
+
+// pciAddressToInt converts the last part of a PCI address string to integer,
+// e.g. 0000:00:16.0 will return 0x16 .
+func pciAddressToInt(pciaddr string) (int, error) {
+	parts := strings.Split(pciaddr, ":")
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("wrong PCI address string '%s': want 3 colon-separated groups, got %d", pciaddr, len(parts))
+	}
+	parts = strings.Split(parts[2], ".")
+	if len(parts) != 2 {
+		return 0, fmt.Errorf("wrong PCI address sub-component '%s': want 2 dot-separated groups, got %d", pciaddr, len(parts))
+	}
+	a, err := strconv.ParseInt(parts[0], 16, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse hex string '%s': %v", parts[0], err)
+	}
+	return int(a), nil
+}
+
+// Open opens an MKHI client connection.
+func (m *MKHIClient) Open(meiPath string) error {
```

Comment:
why not 

OpenMKHIClient(meiPath string) (*MKHIClient, error)

?

Reply:
I would rather use top-level functions for `New*` (not needed here, it's all zero-values) instead than for `Open`, since `Open` could be called at a later stage

Reply:
I don't understand, what later stage? What could be called before Open?

Reply:
You can't call any of the other methods without Open, as far as I can tell.

Reply:
I'm thinking more of scenarios of embedding this struct, where `MKHIClient` is a struct field and Open is called at a later stage with `s.mkhi.Open()` rather than `mei.OpenMKHIClient(s.mkhi)` which feels more natural to use

Reply:
I'm thinking it should be 

mkhi, err := OpenMKHI

s.mkhi = mkhi

As well as

mei, err := OpenMEI

I see this kind of as a poor man's RAII - struct doesn't exist when not initialized, so it cannot be misused when uninitialized